### PR TITLE
[TACHYON-1427] Support setting permission model in Hadoop based TFS

### PR DESCRIPTION
**Pull request by sundapeng**
Jira: https://tachyon.atlassian.net/browse/TACHYON-1427

The pull request is adding support for Hadoop based TFS, change owner, group and permission.

@apc999 , @jsimsa , @yupeng9 , @dongche 

**Since https://github.com/amplab/tachyon/pull/2394 is not merged to branch, I apply my commit TACHYON-1427 based on it. You may need to change to commit view to review my patch. Thanks!**

### CODE REVIEW

**Comment by sundapeng** [comment]
@apc999 , @jsimsa , @yupeng9 , @dongche, I will rebase the pr once https://github.com/amplab/tachyon/pull/2394 is merged. Wait a moment for https://github.com/amplab/tachyon/pull/2394 is also okay.

**Comment by yupeng9** [comment]
Thanks. #2394 is merged. Let us know after you have rebased.

**Comment by sundapeng** [comment]
Rebased with branch  `authorization`

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java [hunk 1]
```diff
@@ -318,11 +319,66 @@ public FileStatus getFileStatus(Path path) throws IOException {
...
+   * @param username username to be set
+   * @param groupname groupname to be set
+   * @throws IOException if the path failed to be changes owner or group
+   */
+  @Override
+  public void setOwner(Path path, final String username, final String groupname)
+      throws IOException {
+    TachyonURI tPath = new TachyonURI(Utils.getPathWithoutScheme(path));
+    Path hdfsPath = Utils.getHDFSPath(tPath, mUnderFSAddress);
+    LOG.info("setOwner({},{},{}) HDFS Path: {} TPath: {}{}", path, username, groupname, hdfsPath,
```
Comment (apc999) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Let's use `LOG.debug`? this is on the critical path.

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java [hunk 2]
```diff
@@ -318,11 +319,66 @@ public FileStatus getFileStatus(Path path) throws IOException {
...
         BLOCK_REPLICATION_CONSTANT, fileStatus.getBlockSizeBytes(), fileStatus.getCreationTimeMs(),
-        fileStatus.getCreationTimeMs(), null, null, null, new Path(mTachyonHeader + tPath));
+            fileStatus.getCreationTimeMs(), new FsPermission((short) fileStatus.getPermission()),
+            fileStatus.getUserName(), fileStatus.getGroupName(), new Path(mTachyonHeader + tPath));
     return ret;
   }
 
   /**
+   * Changes owner or group of a path (i.e. a file or a directory). If username is null, the
+   * original username remains unchanged. Same as groupname.
```
Comment (apc999) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Can you set both? It seems it is allowed. Better to be explicit in the javadoc.

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java [hunk 3]
```diff
@@ -318,11 +319,66 @@ public FileStatus getFileStatus(Path path) throws IOException {
...
+   *
+   * @param path path to set owner or group
+   * @param username username to be set
+   * @param groupname groupname to be set
+   * @throws IOException if the path failed to be changes owner or group
```
Comment (apc999) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: grammar: if changing  owner or group of the path failed.

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java [hunk 4]
```diff
@@ -318,11 +319,66 @@ public FileStatus getFileStatus(Path path) throws IOException {
...
+   * Changes permission of a path.
+   *
+   * @param path path to set permission
+   * @param permission permission set to path
+   * @throws IOException if the path failed to be changed permission
+   */
+  public void setPermission(Path path, FsPermission permission) throws IOException {
+    TachyonURI tPath = new TachyonURI(Utils.getPathWithoutScheme(path));
+    Path hdfsPath = Utils.getHDFSPath(tPath, mUnderFSAddress);
+    LOG.info("setPermission({},{}) HDFS Path: {} TPath : {}{}", path, permission.toString(),
```
Comment (apc999) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: `LOG.debug`

File: integration-tests/src/test/java/tachyon/hadoop/TFSAclIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+import tachyon.Constants;
+import tachyon.LocalTachyonClusterResource;
+import tachyon.security.authentication.AuthType;
+
+/**
+ * Integration tests for {@link TFS#setOwner(Path, String, String)} and
+ * {@link TFS#setPermission(Path, org.apache.hadoop.fs.permission.FsPermission)}.
+ */
+public class TFSAclIntegrationTest {
```
Comment (apc999) [inline integration-tests/src/test/java/tachyon/hadoop/TFSAclIntegrationTest.java]: mark this final class

File: integration-tests/src/test/java/tachyon/hadoop/TFSAclIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    Configuration conf = new Configuration();
+    conf.set("fs.tachyon.impl", TFS.class.getName());
+
+    URI uri = URI.create(sLocalTachyonClusterResource.get().getMasterUri());
+
+    sTFS = FileSystem.get(uri, conf);
+  }
+
+  @Test
+  public void chmodTest() throws Exception {
```
Comment (apc999) [inline integration-tests/src/test/java/tachyon/hadoop/TFSAclIntegrationTest.java]: Please briefly javadoc test cases (you don't need to javadoc exceptions though)

File: integration-tests/src/test/java/tachyon/hadoop/TFSAclIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    create(sTFS, fileA);
+    FileStatus fs = sTFS.getFileStatus(fileA);
+    short permission = fs.getPermission().toShort();
+    // Default permission should be 0644
+    Assert.assertEquals((short) 0644, permission);
+
+    sTFS.setPermission(fileA, FsPermission.createImmutable((short) 0755));
+    Assert.assertEquals((short) 0755, sTFS.getFileStatus(fileA).getPermission().toShort());
+
+    cleanup(sTFS);
```
Comment (apc999) [inline integration-tests/src/test/java/tachyon/hadoop/TFSAclIntegrationTest.java]: move this to `@After` method which will be invoked after each test case?

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java (hunk 4 above)
Comment (apc999) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: TPath -> Tachyon Path, since there is no set term called "TPath"

**Comment by apc999** [comment]
@sundapeng , I left a few comments. Thanks for the patch!

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java (hunk 1 above)
Comment (sundapeng) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Thank you for your comments, I found other methods like `getFileStatus` also use `LOG.info`. How about keep the same log level with other commands?

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java (hunk 2 above)
Comment (sundapeng) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Thanks, I will update the javadoc

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java (hunk 3 above)
Comment (sundapeng) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Thanks,  I will update it.

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java (hunk 4 above)
Comment (sundapeng) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: About LOG.debug
I think we should keep the same log level with other commands.
TPath -> Tachyon Path
Thanks, I will also updated `getFileStatus`

File: integration-tests/src/test/java/tachyon/hadoop/TFSAclIntegrationTest.java (hunk 5 above)
Comment (sundapeng) [inline integration-tests/src/test/java/tachyon/hadoop/TFSAclIntegrationTest.java]: Thanks, I will update it.

File: integration-tests/src/test/java/tachyon/hadoop/TFSAclIntegrationTest.java (hunk 7 above)
Comment (sundapeng) [inline integration-tests/src/test/java/tachyon/hadoop/TFSAclIntegrationTest.java]: Thanks, I will update it.

File: integration-tests/src/test/java/tachyon/hadoop/TFSAclIntegrationTest.java (hunk 6 above)
Comment (sundapeng) [inline integration-tests/src/test/java/tachyon/hadoop/TFSAclIntegrationTest.java]: Thanks, I will update it.

**Comment by sundapeng** [comment]
Patch updated, thank @apc999

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java (hunk 1 above)
Comment (apc999) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: SGTM

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java (hunk 4 above)
Comment (apc999) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: SGTM

**Comment by apc999** [comment]
+1 LGTM.
Thanks @sundapeng

**Comment by apc999** [comment]
@yupeng9  are you OK with the PR ?

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java [hunk 8]
```diff
@@ -318,11 +320,67 @@ public FileStatus getFileStatus(Path path) throws IOException {
...
+   */
+  @Override
+  public void setOwner(Path path, final String username, final String groupname)
+      throws IOException {
+    TachyonURI tPath = new TachyonURI(Utils.getPathWithoutScheme(path));
+    Path hdfsPath = Utils.getHDFSPath(tPath, mUnderFSAddress);
+    LOG.info("setOwner({},{},{}) HDFS Path: {} Tachyon Path: {}{}", path, username, groupname,
+        hdfsPath, mTachyonHeader, tPath);
+    try {
+      if (username != null && !username.isEmpty()) {
```
Comment (jsimsa) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: nit: you could create a single options object and use a single RPC for both username and group permissions.

File: integration-tests/src/test/java/tachyon/hadoop/TFSAclIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+   * Test for {@link TFS#setPermission(Path, org.apache.hadoop.fs.permission.FsPermission)}. It
+   * will test changing the permission of file using TFS.
+   */
+  @Test
+  public void chmodTest() throws Exception {
+    Path fileA = new Path("/chmodfileA");
+
+    create(sTFS, fileA);
+    FileStatus fs = sTFS.getFileStatus(fileA);
+    short permission = fs.getPermission().toShort();
```
Comment (jsimsa) [inline integration-tests/src/test/java/tachyon/hadoop/TFSAclIntegrationTest.java]: nit: no need for a variable if the result is the result is only used in the assert (similar to the 0755 assert).

**Comment by jsimsa** [comment]
@sundapeng thank you for the PR, just two small nits otherwise it LGTM

**Comment by sundapeng** [comment]
Thank @jsimsa for your review, patch updated.

**Comment by apc999** [comment]
+1 LGTM

**Comment by jsimsa** [comment]
LGTM

**Pull request merged**